server/src/main/java/org/elasticsearch/search/lookup/LeafFieldsLookup.java

```diff
@@ -142,16 +142,23 @@ private FieldLookup loadFieldData(String name) {
             cachedFieldData.put(name, data);
         }
         if (data.fields() == null) {
-            String fieldName = data.fieldType().name();
-            fieldVisitor.reset(fieldName);
-            try {
-                reader.document(docId, fieldVisitor);
-                fieldVisitor.postProcess(mapperService);
-                List<Object> storedFields = fieldVisitor.fields().get(data.fieldType().name());
-                data.fields(singletonMap(fieldName, storedFields));
-            } catch (IOException e) {
-                throw new ElasticsearchParseException("failed to load field [{}]", e, name);
+            List<Object> values;
+            if (TypeFieldMapper.NAME.equals(data.fieldType().name())) {
```

Comment:
This approach will not work in 6.x which has to be compatible with indices created in 5.x but I think it is quite sensible in master.